transport/control.go

```diff
@@ -139,6 +172,31 @@ func (qb *quotaPool) add(v int) {
 	}
 }
 
+func (qb *quotaPool) addAndUpdate(v int) {
+	qb.mu.Lock()
+	defer qb.mu.Unlock()
+	qb.version++
+	qb.lockedAdd(v)
+}
+
+func (qb *quotaPool) acquireWithVersion() (<-chan int, uint64) {
+	qb.mu.Lock()
+	version := qb.version
```

Comment:
Maybe an atomic instead of a lock?  You still need the lock for incrementing the version at the same time as the add(), and for compareAndExecute, but that is presumably a more-rare operation than just getting the channel.